File: conformance/dart/events.dart
// Conformance consumer: events sample, Dart target.
//
// Drives the ABI 2 surface end to end: a consumer-implemented `Subscriber`
// callback interface (a Dart class extending the generated abstract class,
// bound through isolate-local NativeCallable vtable trampolines), the
// reference-counted `EventBus` object (constructor, methods, dispose, double
// dispose, use after dispose), an object handed *to* the consumer through a
// callback method (`onAttached` adopts a second `EventBus` wrapper to the
// same native object and keeps it alive past the original's dispose), the
// `Delivery` enum steering `publish`'s accepted count, the buffered
// `Message` record decoded inside a callback, the lazy `Iterable` iterator
// behind `messages()`, the optional `lastMessage()`, the Future-backed
// `publishLater`, and a Dart exception thrown inside a callback surfacing to
// the original caller as `WeaveFFIException` with the foreign code (-4)
// without unwinding through the native frame. Throws (non-zero exit) on any
// mismatch.

import 'package:__PKG__/__LIB__.dart' as wv;

void expect(bool cond, String msg) {
  if (!cond) throw StateError('assertion failed: $msg');
}

/// A subscriber that records every callback it receives. `skipTopic` is
/// routed as [wv.Delivery.skip], `stop` as [wv.Delivery.acceptAndStop], and
/// `failTopic` throws from `route`; `failOnMessage` throws from `onMessage`.
class RecordingSubscriber extends wv.Subscriber {
  RecordingSubscriber(this.name,
      {this.skipTopic = '', this.failTopic = '', this.failOnMessage = false});

  final String name;
  final String skipTopic;
  final String failTopic;
  final bool failOnMessage;

  final List<String> routed = <String>[];
  final List<wv.Message> received = <wv.Message>[];
  int attachedCalls = 0;
  int? subscribersSeenOnAttach;
  wv.EventBus? adoptedBus;

  @override
  wv.Delivery route(String topic) {
    routed.add(topic);
    if (topic == failTopic) throw StateError('$name rejected topic $topic');
    if (topic == skipTopic) return wv.Delivery.skip;
    if (topic == 'stop') return wv.Delivery.acceptAndStop;
    return wv.Delivery.accept;
  }

  @override
  int onMessage(wv.Message message) {
    if (failOnMessage) throw ArgumentError('$name cannot take ${message.text}');
    received.add(message);
    return received.length;
  }

  @override
  void onAttached(wv.EventBus bus) {
    attachedCalls++;
    // The handed-in object is live and usable right here: the producer calls
    // onAttached before it appends this subscriber, so the count excludes us.
    subscribersSeenOnAttach = bus.subscriberCount();
    // The reference is ours. One subscriber keeps it (released at the end of
    // main); the others drop theirs right away.
    if (name == 'keeper') {
      adoptedBus = bus;
    } else {
      bus.dispose();
    }
  }
}

Future<void> main() async {
  // routeOnce: a free function taking the callback interface, whose Delivery
  // return crosses back as the C enum discriminant.
  final probe = RecordingSubscriber('probe', skipTopic: 'quiet');
  expect(wv.routeOnce(probe, 'quiet') == wv.Delivery.skip, 'routeOnce skip');
  expect(wv.routeOnce(probe, 'news') == wv.Delivery.accept, 'routeOnce accept');
  expect(wv.routeOnce(probe, 'stop') == wv.Delivery.acceptAndStop,
      'routeOnce acceptAndStop');
  expect(probe.routed.join(',') == 'quiet,news,stop',
      'routeOnce delivered each topic (got ${probe.routed})');
  expect(probe.attachedCalls == 0, 'routeOnce never attaches');

  // Enum value mapping matches the producer's discriminants.
  expect(wv.Delivery.accept.value == 0 &&
          wv.Delivery.skip.value == 1 &&
          wv.Delivery.acceptAndStop.value == 2,
      'Delivery discriminants');
  expect(wv.Delivery.fromValue(2) == wv.Delivery.acceptAndStop,
      'Delivery.fromValue');

  // A fresh bus: no subscribers, no messages, no last message.
  final bus = wv.EventBus();
  expect(bus.subscriberCount() == 0, 'fresh bus has no subscribers');
  expect(bus.messages().isEmpty, 'fresh bus has no messages');
  expect(bus.lastMessage() == null, 'fresh bus lastMessage is null');

  // subscribe: onAttached fires synchronously with the bus object, and the
  // return value is the new subscriber count.
  final keeper = RecordingSubscriber('keeper', skipTopic: 'quiet');
  final second = RecordingSubscriber('second');
  expect(bus.subscribe(keeper) == 1, 'first subscribe returns 1');
  expect(keeper.attachedCalls == 1, 'keeper attached once');
  expect(keeper.subscribersSeenOnAttach == 0,
      'keeper saw 0 subscribers inside onAttached '
      '(got ${keeper.subscribersSeenOnAttach})');
  expect(keeper.adoptedBus != null, 'keeper adopted the bus reference');
  expect(bus.subscribe(second) == 2, 'second subscribe returns 2');
  expect(second.attachedCalls == 1, 'second attached once');
  expect(second.subscribersSeenOnAttach == 1,
      'second saw 1 subscriber inside onAttached');
  expect(bus.subscriberCount() == 2, 'subscriberCount == 2');

  // The adopted wrapper is a second reference to the same object.
  expect(keeper.adoptedBus!.subscriberCount() == 2,
      'adopted bus sees the same subscribers');

  // publish: both accept -> 2; the Message record decodes inside onMessage
  // with every field (seq, topic, text, tags list).
  expect(bus.publish('news', 'hello', <String>['x', 'y']) == 2,
      'publish news accepted by 2');
  expect(keeper.received.length == 1 && second.received.length == 1,
      'both received news');
  final m = keeper.received.first;
  expect(m.seq == 1, 'first message seq == 1 (got ${m.seq})');
  expect(m.topic == 'news', 'message topic');
  expect(m.text == 'hello', 'message text');
  expect(m.tags.length == 2 && m.tags[0] == 'x' && m.tags[1] == 'y',
      'message tags (got ${m.tags})');
  expect(second.received.first.seq == 1 && second.received.first.text == 'hello',
      'second decoded the same message');

  // Skip: the keeper routes `quiet` as skip, so only `second` accepts.
  expect(bus.publish('quiet', 'psst', <String>[]) == 1, 'publish quiet -> 1');
  expect(keeper.received.length == 1, 'keeper skipped quiet');
  expect(second.received.length == 2 && second.received.last.seq == 2,
      'second received quiet with seq 2');
  expect(second.received.last.tags.isEmpty, 'empty tags list decodes empty');

  // AcceptAndStop: the keeper (first in line) accepts and stops delivery, so
  // `second` never sees it.
  expect(bus.publish('stop', 'last', <String>[]) == 1, 'publish stop -> 1');
  expect(keeper.received.length == 2 && keeper.received.last.text == 'last',
      'keeper received stop');
  expect(second.received.length == 2, 'second not reached after stop');
  expect(keeper.routed.join(',') == 'news,quiet,stop', 'keeper routing log');
  expect(second.routed.join(',') == 'news,quiet', 'second routing log');

  // Iterator: lazy Iterable over the native iterator, in publish order; a
  // second iteration launches a fresh native iterator, and an abandoned
  // partial iteration is harmless.
  final texts = bus.messages().toList();
  expect(texts.join('|') == 'hello|psst|last',
      'messages in order (got $texts)');
  expect(bus.messages().length == 3, 'second iteration yields 3 again');
  expect(bus.messages().first == 'hello', 'partial iteration');
  expect(bus.messages().skip(1).first == 'psst', 'skip then first');

  // Optional record return.
  final last = bus.lastMessage();
  expect(last != null, 'lastMessage present');
  expect(last!.seq == 3 && last.topic == 'stop' && last.text == 'last',
      'lastMessage is the stop message');

  // Foreign error from route: a Dart exception inside the callback aborts the
  // publish and surfaces to the caller as the generic exception carrying the
  // foreign code and the exception text; the VM does not crash, and the bus
  // stays usable. Earlier subscribers already accepted the message.
  final rejecting = RecordingSubscriber('rejecting', failTopic: 'boom');
  expect(bus.subscribe(rejecting) == 3, 'third subscribe returns 3');
  try {
    bus.publish('boom', 'x', <String>[]);
    throw StateError('expected WeaveFFIException for rejected topic');
  } on wv.WeaveFFIException catch (e) {
    expect(e.code == wv.WeaveFFIException.foreignCode,
        'foreign code -4 (got ${e.code})');
    expect(e.code == -4, 'foreignCode constant is -4');
    expect(e.message.contains('rejecting rejected topic boom'),
        'foreign message carries the Dart exception text (got ${e.message})');
  }
  expect(keeper.received.length == 3 && second.received.length == 3,
      'subscribers ahead of the failing one still accepted boom');
  expect(rejecting.received.isEmpty, 'rejecting never received');
  expect(bus.messages().length == 4, 'the aborted publish was still logged');
  expect(bus.lastMessage()!.text == 'x', 'lastMessage is the aborted one');
  expect(bus.publish('ok', 'y', <String>['t']) == 3,
      'bus usable after a foreign error');
  expect(rejecting.received.length == 1 && rejecting.received.first.seq == 5,
      'rejecting received ok with seq 5');

  // Foreign error from onMessage (a different vtable slot and exception type).
  final fragile = RecordingSubscriber('fragile', failOnMessage: true);
  expect(bus.subscribe(fragile) == 4, 'fourth subscribe returns 4');
  try {
    bus.publish('any', 'payload', <String>[]);
    throw StateError('expected WeaveFFIException for failing onMessage');
  } on wv.WeaveFFIException catch (e) {
    expect(e.code == wv.WeaveFFIException.foreignCode,
        'onMessage foreign code -4 (got ${e.code})');
    expect(e.message.contains('fragile cannot take payload'),
        'onMessage foreign message (got ${e.message})');
  }
  expect(fragile.routed.length == 1, 'fragile was routed once');
  expect(bus.subscriberCount() == 4, 'a failing subscriber stays subscribed');

  // Foreign error from onAttached: subscribe itself fails and the subscriber
  // is not retained.
  final unattachable = FailingAttachSubscriber();
  try {
    bus.subscribe(unattachable);
    throw StateError('expected WeaveFFIException for failing onAttached');
  } on wv.WeaveFFIException catch (e) {
    expect(e.code == -4, 'onAttached foreign code -4 (got ${e.code})');
    expect(e.message.contains('refuses to attach'),
        'onAttached foreign message (got ${e.message})');
  }
  expect(bus.subscriberCount() == 4, 'failed subscribe did not retain');

  // clearSubscribers drops every retained subscriber; publishing then
  // reaches nobody but is still logged.
  bus.clearSubscribers();
  expect(bus.subscriberCount() == 0, 'cleared subscribers');
  expect(bus.publish('news', 'nobody', <String>[]) == 0,
      'publish with no subscribers -> 0');
  expect(keeper.received.length == 5,
      'cleared keeper receives nothing (got ${keeper.received.length})');

  // Async: publishLater runs on a producer thread and settles a Future. The
  // generated Dart callback trampolines are isolate-local, so subscribers
  // can't be driven from that thread; with none attached the call resolves
  // with 0 accepted and the message is logged.
  final laterCount = await bus.publishLater('later', 'zzz');
  expect(laterCount == 0, 'publishLater resolves with 0 (got $laterCount)');
  expect(bus.lastMessage()!.text == 'zzz' && bus.lastMessage()!.seq == 8,
      'publishLater logged its message (got ${bus.lastMessage()!.text})');
  expect(bus.messages().length == 8, 'eight messages logged');
  final second1 = await bus.publishLater('later', 'again');
  expect(second1 == 0 && bus.messages().length == 9, 'second publishLater');

  // Object lifetime: dispose the original wrapper; the adopted reference
  // keeps the native bus alive and fully usable. Double dispose is safe and
  // use after dispose is a StateError rather than a native fault.
  bus.dispose();
  bus.dispose();
  try {
    bus.subscriberCount();
    throw StateError('expected StateError after dispose');
  } on StateError catch (e) {
    expect(e.message.contains('dispose'), 'use after dispose message');
  }
  final adopted = keeper.adoptedBus!;
  expect(adopted.messages().length == 9,
      'adopted reference outlives the original wrapper');
  expect(adopted.publish('final', 'bye', <String>[]) == 0,
      'publish through the adopted reference');
  expect(adopted.lastMessage()!.text == 'bye', 'adopted lastMessage');
  adopted.dispose();
  adopted.dispose();

  // A bus that is dropped while still holding subscribers releases them on
  // its last reference (the consumer's `free` entries run through a listener
  // NativeCallable once the event loop turns).
  final short = wv.EventBus();
  final tenant = RecordingSubscriber('tenant');
  short.subscribe(tenant);
  tenant.adoptedBus?.dispose();
  short.dispose();
  await Future<void>.delayed(Duration.zero);

  print('dart/events: OK');
}

/// A subscriber whose `onAttached` throws, so `subscribe` itself fails.
class FailingAttachSubscriber extends wv.Subscriber {
  @override
  wv.Delivery route(String topic) => wv.Delivery.accept;

  @override
  int onMessage(wv.Message message) => 0;

  @override
  void onAttached(wv.EventBus bus) {
    bus.dispose();
    throw StateError('refuses to attach');
  }
}

File: conformance/kotlin/events.kt
// Conformance consumer: events sample, Kotlin (JVM via JNI) target.
//
// Exercises the ABI 2 callback-interface and object surface: `Subscriber` is
// a generated Kotlin interface the consumer implements (the JNI vtable
// trampolines pin the implementing object with a GlobalRef and dispatch
// through the `SubscriberJni` shims), `EventBus` is a reference-counted
// `AutoCloseable` wrapper (companion `invoke` for `new`, `close()` releases
// one strong reference, the `Cleaner` is the backstop), and `Message` is a
// data class decoded from a value buffer. Asserts that every callback method
// is invoked with the right arguments (including the bus object handed to
// `onAttached`, which is usable and independently closeable), that `Delivery`
// return values steer `publish`'s accepted count, that a Kotlin exception
// thrown inside a callback surfaces to the caller as `WeaveFFIException` with
// code -4 without unwinding through the JVM, the `suspend` async `publishLater`
// driven with `runBlocking`, the iterator-backed `messages()` drained as a
// `Sequence`, the nullable `lastMessage()`, `routeOnce` without a bus, and the
// close semantics (double `close()` is safe, use after close throws).
// Compiled in-module with the generated `WeaveFFI.kt`.
@file:JvmName("Main")

import com.weaveffi.Delivery
import com.weaveffi.EventBus
import com.weaveffi.EventsEventBusMessagesIterator
import com.weaveffi.Message
import com.weaveffi.Subscriber
import com.weaveffi.WeaveFFI
import com.weaveffi.WeaveFFIException
import java.lang.ref.WeakReference
import kotlin.system.exitProcess
import kotlinx.coroutines.runBlocking

fun expect(cond: Boolean, msg: String) {
    if (!cond) {
        System.err.println("assertion failed: $msg")
        exitProcess(1)
    }
}

/** Run `block` and return the exception it threw, or null if it completed. */
inline fun thrownBy(block: () -> Unit): Throwable? =
    try {
        block()
        null
    } catch (e: Throwable) {
        e
    }

/** Spin the collector until `ref` clears or we give up. */
fun collected(ref: WeakReference<*>): Boolean {
    for (i in 0 until 200) {
        if (ref.get() == null) return true
        System.gc()
        System.runFinalization()
        Thread.sleep(5)
    }
    return ref.get() == null
}

/**
 * A subscriber that records every callback. `skipTopic` is routed as Skip,
 * `stopTopic` as AcceptAndStop, `failTopic` throws from `route`, and
 * `failOnMessage` throws from `onMessage` instead.
 */
class RecordingSubscriber(
    val name: String,
    private val skipTopic: String? = null,
    private val stopTopic: String? = null,
    private val failTopic: String? = null,
    private val failOnMessage: Boolean = false,
    private val failOnAttached: Boolean = false,
) : Subscriber {
    val routed = mutableListOf<String>()
    val messages = mutableListOf<Message>()
    var attachedCount = 0
    var subscriberCountAtAttach = -1L
    var keptBus: EventBus? = null

    override fun route(topic: String): Delivery {
        routed.add(topic)
        if (topic == failTopic) throw IllegalStateException("subscriber $name rejected $topic")
        return when (topic) {
            skipTopic -> Delivery.Skip
            stopTopic -> Delivery.AcceptAndStop
            else -> Delivery.Accept
        }
    }

    override fun onMessage(message: Message): Long {
        if (failOnMessage) throw RuntimeException("subscriber $name failed on message ${message.seq}")
        messages.add(message)
        return messages.size.toLong()
    }

    override fun onAttached(bus: EventBus) {
        attachedCount++
        if (failOnAttached) throw IllegalArgumentException("subscriber $name refused to attach")
        // The bus arrives as an adopted strong reference: it's usable right
        // here (subscribe hands it over before it takes its lock) and we keep
        // the wrapper so a later close proves it's independent of the caller's.
        subscriberCountAtAttach = bus.subscriberCount()
        keptBus = bus
    }
}

/** Route once through a subscriber nothing else references, returning a weak handle to it. */
fun routeOnceThrowaway(): WeakReference<RecordingSubscriber> {
    val sub = RecordingSubscriber("throwaway")
    expect(WeaveFFI.routeOnce(sub, "t") == Delivery.Accept, "routeOnce throwaway")
    return WeakReference(sub)
}

/** Subscribe a subscriber nothing else references, returning a weak handle to it. */
fun subscribeThrowaway(bus: EventBus): WeakReference<RecordingSubscriber> {
    val sub = RecordingSubscriber("gone")
    expect(bus.subscribe(sub) == 4L, "subscribe gone")
    sub.keptBus!!.close()
    return WeakReference(sub)
}

fun main() {
    val bus = EventBus()
    expect(bus.subscriberCount() == 0L, "fresh bus has no subscribers")
    expect(bus.lastMessage() == null, "fresh bus has no last message")
    expect(bus.messages().asSequence().toList().isEmpty(), "fresh bus has no messages")
    expect(bus.publish("void", "nobody home", listOf()) == 0L, "publish with no subscribers accepts 0")

    // Subscribe three subscribers; each gets exactly one onAttached with the
    // bus, observing the count before it was added.
    val quiet = RecordingSubscriber("quiet", skipTopic = "quiet")
    val stopper = RecordingSubscriber("stopper", stopTopic = "stop")
    val tail = RecordingSubscriber("tail")
    expect(bus.subscribe(quiet) == 1L, "first subscribe returns 1")
    expect(bus.subscribe(stopper) == 2L, "second subscribe returns 2")
    expect(bus.subscribe(tail) == 3L, "third subscribe returns 3")
    expect(bus.subscriberCount() == 3L, "subscriberCount == 3")
    for ((i, s) in listOf(quiet, stopper, tail).withIndex()) {
        expect(s.attachedCount == 1, "${s.name} attached once (got ${s.attachedCount})")
        expect(
            s.subscriberCountAtAttach == i.toLong(),
            "${s.name} saw $i subscribers at attach (got ${s.subscriberCountAtAttach})"
        )
        expect(s.keptBus != null, "${s.name} kept the bus")
    }

    // Delivery steering: Accept everywhere -> 3; Skip for `quiet` -> 2;
    // AcceptAndStop in the middle -> 2 and `tail` never sees it.
    expect(bus.publish("news", "hello", listOf("a", "b")) == 2L + 1L, "publish news accepted by 3")
    expect(bus.publish("quiet", "psst", listOf()) == 2L, "publish quiet accepted by 2")
    expect(bus.publish("stop", "last", listOf("z")) == 2L, "publish stop accepted by 2")
    expect(quiet.routed == listOf("news", "quiet", "stop"), "quiet routed every topic (got ${quiet.routed})")
    expect(stopper.routed == listOf("news", "quiet", "stop"), "stopper routed every topic")
    expect(tail.routed == listOf("news", "quiet"), "tail never routed stop (got ${tail.routed})")
    expect(quiet.messages.map { it.text } == listOf("hello", "last"), "quiet got hello,last (got ${quiet.messages.map { it.text }})")
    expect(stopper.messages.map { it.text } == listOf("hello", "psst", "last"), "stopper got all three")
    expect(tail.messages.map { it.text } == listOf("hello", "psst"), "tail got hello,psst (got ${tail.messages.map { it.text }})")

    // Message record fields decoded from the callback's buffer. seq counts
    // from 1 and includes the subscriber-less publish above.
    val first = stopper.messages[0]
    expect(first == Message(2L, "news", "hello", listOf("a", "b")), "first message record (got $first)")
    expect(stopper.messages[1].tags.isEmpty(), "psst carries no tags")
    expect(stopper.messages[2] == Message(4L, "stop", "last", listOf("z")), "last message record (got ${stopper.messages[2]})")

    // Async publish via the suspend wrapper, resumed from the producer thread.
    val later = runBlocking { bus.publishLater("later", "async hello") }
    expect(later == 3L, "publishLater accepted by 3 (got $later)")
    expect(stopper.messages.last().text == "async hello", "async message delivered")
    expect(stopper.messages.last().tags.isEmpty(), "async message has no tags")
    expect(stopper.messages.last().seq == 5L, "async message seq 5 (got ${stopper.messages.last().seq})")

    // Iterator-backed messages(): drained as a Sequence, and again by hand.
    val texts = bus.messages().asSequence().toList()
    expect(
        texts == listOf("nobody home", "hello", "psst", "last", "async hello"),
        "messages in publish order (got $texts)"
    )
    val byHand = mutableListOf<String>()
    val it = bus.messages()
    while (it.hasNext()) byHand.add(it.next())
    expect(byHand == texts, "manual iteration matches")
    expect(thrownBy { it.next() } is NoSuchElementException, "exhausted iterator throws NoSuchElementException")
    // An abandoned iterator can be closed early, and closed twice.
    val partial = bus.messages() as EventsEventBusMessagesIterator
    expect(partial.next() == "nobody home", "partial iterator first element")
    partial.close()
    partial.close()
    expect(!partial.hasNext(), "closed iterator reports no more elements")

    // Nullable record return.
    val last = bus.lastMessage()
    expect(last == Message(5L, "later", "async hello", listOf()), "lastMessage (got $last)")

    // routeOnce: a free function taking the callback interface without a bus.
    expect(WeaveFFI.routeOnce(quiet, "quiet") == Delivery.Skip, "routeOnce quiet -> Skip")
    expect(WeaveFFI.routeOnce(stopper, "stop") == Delivery.AcceptAndStop, "routeOnce stop -> AcceptAndStop")
    expect(WeaveFFI.routeOnce(tail, "anything") == Delivery.Accept, "routeOnce -> Accept")
    expect(quiet.routed.last() == "quiet", "routeOnce reached route()")
    // The producer drops its only reference when routeOnce returns, so the
    // GlobalRef pinning a throwaway subscriber is released and it becomes
    // collectable once the frame that created it is gone.
    expect(collected(routeOnceThrowaway()), "routeOnce released its subscriber (free ran)")

    // The bus handed to onAttached is a distinct wrapper over the same
    // object: usable, and closing it leaves the caller's wrapper alive.
    val kept = tail.keptBus!!
    expect(kept.handle == bus.handle, "onAttached bus is the same native object")
    expect(kept.subscriberCount() == 3L, "kept bus reads the live subscriber count")
    kept.close()
    kept.close()
    expect(thrownBy { kept.subscriberCount() } is IllegalStateException, "kept bus rejects use after close")
    expect(bus.subscriberCount() == 3L, "original bus still alive after closing the kept wrapper")
    quiet.keptBus!!.close()
    stopper.keptBus!!.close()

    // Foreign errors: a Kotlin exception thrown from any callback method
    // surfaces to the caller as WeaveFFIException(-4) carrying the throwable's
    // text, and the JVM keeps running. Use a fresh bus for exact counts.
    EventBus().use { bus2 ->
        val ok = RecordingSubscriber("ok")
        val rejecter = RecordingSubscriber("rejecter", failTopic = "boom")
        bus2.subscribe(ok)
        bus2.subscribe(rejecter)
        ok.keptBus!!.close()
        rejecter.keptBus!!.close()
        val routeErr = thrownBy { bus2.publish("boom", "x", listOf()) }
        expect(routeErr is WeaveFFIException, "throwing route() surfaces as WeaveFFIException (got $routeErr)")
        expect((routeErr as WeaveFFIException).code == -4, "foreign error code -4 (got ${routeErr.code})")
        expect(
            routeErr.message?.contains("rejected boom") == true,
            "foreign error carries the Kotlin message (got ${routeErr.message})"
        )
        expect(routeErr.message?.contains("IllegalStateException") == true, "foreign error names the exception class")
        // The earlier subscriber was still delivered to, the producer logged
        // the message, and the bus remains usable.
        expect(ok.messages.map { it.text } == listOf("x"), "subscriber before the failure still received x")
        expect(bus2.lastMessage()?.text == "x", "producer logged the message despite the failure")
        expect(bus2.publish("fine", "y", listOf()) == 2L, "bus usable after a foreign error")

        val onMsg = RecordingSubscriber("onMsg", failOnMessage = true)
        bus2.subscribe(onMsg)
        onMsg.keptBus!!.close()
        val msgErr = thrownBy { bus2.publish("any", "z", listOf()) }
        expect(msgErr is WeaveFFIException && msgErr.code == -4, "throwing onMessage() surfaces as -4 (got $msgErr)")
        expect(msgErr?.message?.contains("failed on message") == true, "onMessage error text (got ${msgErr?.message})")

        // Throwing from onAttached aborts subscribe itself and the subscriber
        // isn't retained.
        val refuser = RecordingSubscriber("refuser", failOnAttached = true)
        val attachErr = thrownBy { bus2.subscribe(refuser) }
        expect(attachErr is WeaveFFIException && attachErr.code == -4, "throwing onAttached() surfaces as -4 (got $attachErr)")
        expect(bus2.subscriberCount() == 3L, "refused subscriber not retained (got ${bus2.subscriberCount()})")

        // routeOnce also propagates the foreign error.
        val onceErr = thrownBy { WeaveFFI.routeOnce(rejecter, "boom") }
        expect(onceErr is WeaveFFIException && onceErr.code == -4, "routeOnce propagates -4 (got $onceErr)")
        expect(WeaveFFI.routeOnce(rejecter, "calm") == Delivery.Accept, "rejecter still works for other topics")
    }

    // clearSubscribers drops every retained subscriber (each free runs), and
    // the bus keeps its log.
    val goneRef = subscribeThrowaway(bus)
    expect(bus.subscriberCount() == 4L, "subscribe gone")
    bus.clearSubscribers()
    expect(bus.subscriberCount() == 0L, "clearSubscribers empties the bus")
    expect(collected(goneRef), "cleared subscriber was released by the producer")
    expect(bus.publish("after", "nobody", listOf()) == 0L, "publish after clear accepts 0")
    expect(bus.lastMessage()?.text == "nobody", "log continues after clear")
    expect(quiet.messages.size == 3, "cleared subscribers receive nothing more")

    // Close semantics: double close is safe, use after close throws.
    bus.close()
    bus.close()
    expect(thrownBy { bus.subscriberCount() } is IllegalStateException, "bus rejects use after close")
    expect(thrownBy { bus.publish("x", "y", listOf()) } is IllegalStateException, "publish rejects use after close")

    println("kotlin/events: OK")
}
